mutex.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package mutex
+
+import (
+	"regexp"
+	"time"
+
+	"github.com/juju/errors"
+)
+
+var (
+	validName = regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9-]*$")
+)
+
+// Releaser defines the Release method that is the only thing that can be done
+// to a acquired mutex.
+type Releaser interface {
+	// Release makes the mutex available for others and invalidates the instance.
+	// Subsequent calls to Release will error.
+	Release() error
```

Comment:
I'm trying to remember a time where I've thought "golly, that panic was a good idea, I'm glad it took down the controller" rather than "GAAAAH FFS". Inability to release a global lock it holds probably does qualify; and if the client was doing something super-evil that we somehow couldn't prevent, I might appreciate that; but a double-release (that we can trivially detect and deflect *without* panicking) is entirely predictable.

(counterargument: if we don't impose extreme consequences for double-releasing, people will just do it anyway and ignore all errors... and, well, yeah, they *could*, but that would just be shitty code, which I think is (sadly) unexceptional and (happily) at least addressable.)